### PR TITLE
Move diagnostics out from QueryJob and optimize for the case with no diagnostics

### PR DESCRIPTION
**Pull request by Zoxc**
r? @michaelwoerister

### CODE REVIEW

**Comment by Zoxc** [comment]
@rust-timer build 203d99c0ed5ec7c17b4cab82a3995a30de4903ce

File: src/librustc/ty/query/plumbing.rs [hunk 1]
```diff
@@ -221,13 +224,19 @@ impl<'a, 'tcx, Q: QueryDescription<'tcx>> JobOwner<'a, 'tcx, Q> {
             tls::enter_context(&new_icx, |_| {
                 compute(tcx)
             })
-        });
+        })
+    }
 
-        // Extract the diagnostic from the job
-        let diagnostics = mem::replace(&mut *self.job.diagnostics.lock(), Vec::new());
```
Comment (Zoxc) [inline src/librustc/ty/query/plumbing.rs]: Note: Another lock gone here too.

**Comment by pietroalbini** [comment]
@bors retry
AppVeyor... what's wrong with you today?

**Pull request merged**